src/NuGet.Clients/PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -38,28 +38,142 @@ public UIActionEngine(ISourceRepositoryProvider sourceProvider, NuGetPackageMana
+        public async Task PerformUpdateAsync(
+            INuGetUI uiService,
+            List<PackageIdentity> packagesToUpdate,
+            DependencyObject windowOwner,
+            CancellationToken token)
+        {
+            await PerformActionImplAsync(
+                uiService,
+                () =>
+                {
+                    return ResolveActionsForUpdate(
+                        uiService,
+                        packagesToUpdate,
+                        token);
+                },
+                async (actions) =>
+                {
+                    foreach (var projectActions in actions.GroupBy(action => action.Item1))
+                    {
+                        await _packageManager.ExecuteNuGetProjectActionsAsync(
+                           projectActions.Key,
+                           projectActions.Select(action => action.Item2),
+                           uiService.ProgressWindow,
+                           token);
+                    }
+                },
+                windowOwner,
+                token);
+        }
+
+        /// <summary>
+        /// Calculates the list of actions needed to perform packages updates.
+        /// </summary>
+        /// <param name="uiService">ui service.</param>
+        /// <param name="packagesToUpdate">The list of packages to update.</param>
+        /// <param name="token">Cancellation token.</param>
+        /// <returns>The list of actions.</returns>
+        private async Task<IReadOnlyList<Tuple<NuGetProject, NuGetProjectAction>>> ResolveActionsForUpdate(
+            INuGetUI uiService,
+            List<PackageIdentity> packagesToUpdate,
+            CancellationToken token)
+        {
+            var resolvedActions = new List<Tuple<NuGetProject, NuGetProjectAction>>();
+
+            foreach (var project in uiService.Projects)
             {
-                uiService.ShowProgressDialog(windowOwner);
+                var installedPackages = await project.GetInstalledPackagesAsync(token);
+                HashSet<string> packageIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                foreach (var p in installedPackages)
+                {
+                    packageIds.Add(p.PackageIdentity.Id);
+                }
+                var packagesToUpdateInProject = packagesToUpdate.Where(
+                    package => packageIds.Contains(package.Id)).ToList();
 
-                var projects = uiService.Projects;
+                if (packagesToUpdateInProject.Any())
+                {
+                    var includePrerelease = packagesToUpdateInProject.Where(
+                        package => package.Version.IsPrerelease).Any();
+
+                    var resolutionContext = new ResolutionContext(
+                        uiService.DependencyBehavior,
+                        includePrelease: includePrerelease,
+                        includeUnlisted: false,
```

Comment:
During uninstall, You have done 'var includeUnlisted = userAction.Action == NuGetProjectActionType.Uninstall;'.
Why? And, Why is it okay to have unlisted = false during update? Because, you might be uninstalling packages.